buildSrc/src/main/kotlin/org/gradle/plugins/buildscan/BuildScanConfigurationPlugin.kt

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.plugins.buildscan
+
+import com.gradle.scan.plugin.BuildScanExtension
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.quality.Checkstyle
+import org.gradle.api.plugins.quality.CodeNarc
+import org.gradle.api.reporting.ReportContainer
+import org.gradle.api.reporting.Reporting
+import org.gradle.kotlin.dsl.the
+import org.jsoup.Jsoup
+import org.jsoup.parser.Parser
+import org.gradle.api.internal.GradleInternal
+import org.gradle.internal.classloader.ClassLoaderHierarchyHasher
+import kotlin.concurrent.thread
+
+
+class BuildScanConfigurationPlugin : Plugin<Project> {
+    private
+    val isCiServer = System.getenv().containsKey("CI")
+
+    override fun apply(project: Project): Unit = project.run {
+        apply {
+            plugin("com.gradle.build-scan")
```

Comment:
I think it would be nice to keep the lazy logic like before:
```
pluginManager.withPlugin("com.gradle.build-scan") {
...
}
```

Why can't or shouldn't we do this anymore?

Reply:
There is no point in it anymore because the way we need to workaround the auto-apply issue. The sheer presence of the build-scan-plugin jar in the classpath (added by buildSrc) leads to an exception if plugin is auto-applied via `--scan`. The only way to avoid this exception at this point, is to apply the build-scan plugin in the root project via the new plugin DSL. As it is now always applied, there is no point in the lazyness anymore. I think it is not too much of a deal, as we expect it to be always applied anyway. We would need to rework the auto-apply mechanisms to avoid this. Something to be considered.